pkg/auth/auth-result.go

```diff
@@ -0,0 +1,71 @@
+package auth
+
+import (
+	"net/http"
+)
+
+type resultType int
+
+const (
+	unknown resultType = iota
+	forbidden
+	unauthorized
+	authorized
+)
+
+type header struct {
+	key   string
+	value string
+}
+
+func Header(key string, value string) header {
+	return header{key, value}
+}
+
+type AuthResult struct {
+	resultType resultType
+	headers    []header
+}
+
+func (a *AuthResult) WriteResponse(w http.ResponseWriter) bool {
+	switch a.resultType {
+	case authorized:
+		// Do not write a response, we're authorized
+		return false
+	case unauthorized:
+		for _, h := range a.headers {
+			w.Header().Add(h.key, h.value)
+		}
```

Comment:
The `Header()` structure is actually a `map` with `string` keys and `[]string` values. I'm not 100% sure, but I think (based on [this](https://github.com/golang/go/blob/945a2b17f3b47b4e0bda9a2a92186412b5704c9a/src/net/http/header.go#L173-L187)) the headers will be first sorted alphabetically, then in order of addition (for multiples of the same header).

Reply:
> I'm not 100% sure, but I think (based on [this](https://github.com/golang/go/blob/945a2b17f3b47b4e0bda9a2a92186412b5704c9a/src/net/http/header.go#L173-L187)) the headers will be first sorted alphabetically, then in order of addition (for multiples of the same header).

The order of (repeated) auth headers matters.

```text
WWW-Authenticate: Bearer authorize="idp.example.com/oauth"
WWW-Authenticate: Basic realm="example.com"
```

..may be treated differently by clients compared to..

```text
WWW-Authenticate: Basic realm="example.com"
WWW-Authenticate: Bearer authorize="idp.example.com/oauth"
```

Of course if order of repeated headers matters, the middleware could always embed multiple header values in the same field:

```text
WWW-Authenticate: Bearer authorize="idp.example.com/oauth", Basic realm="example.com"
```

If it's possible, maintaining the order the middleware gave would probably be the least surprising behaviour.

Reply:
Repeated headers will be kept in the order that they're provided by the middleware (they're added to `Header()` in order [here](https://github.com/git-ecosystem/git-bundle-server/blob/40a02483218e97df418ce868f4230e2722173dc4/pkg/auth/auth-result.go#L36-L38), then added to the response in order [here](https://github.com/golang/go/blob/945a2b17f3b47b4e0bda9a2a92186412b5704c9a/src/net/http/header.go#L211-L223).

Since the `AuthResult` methods will be in a public package, I'll also add formal function documentation to make that clear.

Reply:
> Repeated headers will be kept in the order that they're provided by the middleware

Ah, then that should be fine!